File: tools/devserver/units.py
# minimoni - zero-dependency system monitoring
# Copyright (C) 2026 Javier Beaumont <javierbeaumont@users.noreply.github.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

"""
Unit conversions for the minimoni dev server, mirroring src/http.c exactly.

Raw base-unit metrics (from mock_data) are converted to the configured display
units: /api/current uses the *_card_unit values, /api/metrics uses *_chart_unit.
Memory/disk raw fields are OMITTED when the unit is "%" (as the real server does,
so the dashboard falls back to the percentage), and the load is normalised by
core count when its unit is "%".
"""

from typing import cast

from mock_data import JSON

# The real server reads the core count from sysfs; the dashboard assumes 4 when
# normalising load client-side, so the dev server uses the same assumption.
CORES = 4


def net_convert(bps: float, unit: str) -> float:
    if not unit or unit[0] == "m":
        if len(unit) > 2 and unit[1] == "b" and unit[2] == "p":  # mbps
            return bps * 8.0 / 1e6
        return bps / 1048576.0  # mb
    if unit[0] == "g":
        if len(unit) > 2 and unit[1] == "b" and unit[2] == "p":  # gbps
            return bps * 8.0 / 1e9
        return bps / 1073741824.0  # gb
    if unit[0] == "k":
        if len(unit) > 2 and unit[1] == "b" and unit[2] == "p":  # kbps
            return bps * 8.0 / 1000.0
        return bps / 1024.0  # kb
    return bps / 1048576.0


def mem_convert(mb: float, unit: str) -> float:
    return mb / 1024.0 if unit and unit[0] == "g" else mb


def disk_convert(gb: float, unit: str) -> float:
    return gb / 1024.0 if unit and unit[0] == "t" else gb


def temp_convert(celsius: float, unit: str, ref: float) -> float:
    if not unit:
        return celsius
    if unit[0] == "f":
        return celsius * 9.0 / 5.0 + 32.0
    if unit[0] == "%":
        return celsius * 100.0 / ref if ref > 0 else celsius
    return celsius


def temp_ref(crit: float | None, fallback: float) -> float:
    """100% reference for temp percent mode: the sysfs critical trip point when
    present, else the configured fallback. Mirrors the C server."""
    return crit if crit is not None else fallback


def load_convert(load: float, cores: int, unit: str) -> float:
    if unit and unit[0] == "%" and cores > 0:
        return load * 100.0 / cores
    return load


def to_current(raw: JSON, f: JSON, fallback: float) -> JSON:
    """Serialize a raw snapshot for /api/current using the CARD units."""
    # The metric values are numeric at runtime; the str/int passthroughs
    # (timestamp, uptime) are read from `raw` directly.
    m = cast(dict[str, float], raw)
    lu, mu = str(f["cpu_load_card_unit"]), str(f["mem_card_unit"])
    du, nu, tu = (
        str(f["disk_card_unit"]),
        str(f["net_card_unit"]),
        str(f["temp_card_unit"]),
    )

    out: JSON = {
        "timestamp": raw["timestamp"],
        "load_1m": round(load_convert(m["load_1m"], CORES, lu), 2),
        "load_5m": round(load_convert(m["load_5m"], CORES, lu), 2),
        "load_15m": round(load_convert(m["load_15m"], CORES, lu), 2),
        "cpu_user_percent": round(m["cpu_user"], 1),
        "cpu_system_percent": round(m["cpu_system"], 1),
        "cpu_idle_percent": round(m["cpu_idle"], 1),
    }
    if mu[0] != "%":
        out["mem_used"] = round(mem_convert(m["mem_used_mb"], mu), 2)
        out["mem_available"] = round(mem_convert(m["mem_avail_mb"], mu), 2)
        out["mem_total"] = round(mem_convert(m["mem_total_mb"], mu), 2)
    out["mem_percent"] = round(m["mem_percent"], 1)
    if du[0] != "%":
        out["disk_used"] = round(disk_convert(m["disk_used_gb"], du), 2)
        out["disk_total"] = round(disk_convert(m["disk_total_gb"], du), 2)
        out["disk_free"] = round(disk_convert(m["disk_free_gb"], du), 2)
    out["disk_percent"] = round(m["disk_percent"], 1)

    tc = m.get("temp_c")
    crit = m.get("temp_critical_c")
    ref = temp_ref(crit, fallback)
    out["temp"] = round(temp_convert(tc, tu, ref), 1) if tc is not None else None
    out["temp_critical"] = (
        round(temp_convert(crit, tu, ref), 1) if crit is not None else None
    )

    out["net_rx"] = round(net_convert(m["net_rx_bps"], nu), 2)
    out["net_tx"] = round(net_convert(m["net_tx_bps"], nu), 2)
    out["uptime_seconds"] = raw["uptime"]
    return out


def to_point(raw: JSON, f: JSON, fallback: float) -> JSON:
    """Serialize a raw point for /api/metrics using the CHART units (short keys)."""
    # The metric values are numeric at runtime; the int passthrough (t, uptime)
    # is read from `raw` directly.
    m = cast(dict[str, float], raw)
    lu, mu = str(f["cpu_load_chart_unit"]), str(f["mem_chart_unit"])
    du, nu, tu = (
        str(f["disk_chart_unit"]),
        str(f["net_chart_unit"]),
        str(f["temp_chart_unit"]),
    )

    out: JSON = {
        "t": raw["t"],
        "l1": round(load_convert(m["load_1m"], CORES, lu), 2),
        "l5": round(load_convert(m["load_5m"], CORES, lu), 2),
        "l15": round(load_convert(m["load_15m"], CORES, lu), 2),
        "cu": round(m["cpu_user"], 1),
        "cs": round(m["cpu_system"], 1),
        "ci": round(m["cpu_idle"], 1),
    }
    if mu[0] != "%":
        out["mu"] = round(mem_convert(m["mem_used_mb"], mu), 2)
        out["ma"] = round(mem_convert(m["mem_avail_mb"], mu), 2)
        out["mt"] = round(mem_convert(m["mem_total_mb"], mu), 2)
    out["mp"] = round(m["mem_percent"], 1)
    if du[0] != "%":
        out["du"] = round(disk_convert(m["disk_used_gb"], du), 2)
        out["dt"] = round(disk_convert(m["disk_total_gb"], du), 2)
        out["df"] = round(disk_convert(m["disk_free_gb"], du), 2)
    out["dp"] = round(m["disk_percent"], 1)

    tc = m.get("temp_c")
    ref = temp_ref(m.get("temp_critical_c"), fallback)
    out["tp"] = round(temp_convert(tc, tu, ref), 1) if tc is not None else None

    out["nr"] = round(net_convert(m["net_rx_bps"], nu), 2)
    out["nt"] = round(net_convert(m["net_tx_bps"], nu), 2)
    out["up"] = raw["uptime"]
    return out

File: tools/devserver/units.js
/*
 * minimoni - zero-dependency system monitoring
 * Copyright (C) 2026 Javier Beaumont <javierbeaumont@users.noreply.github.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

/* Unit conversions + /api serialization, mirroring the C server (src/units.c,
 * src/json.c, src/http.c). */

'use strict';

/* The mock has no real CPU, so it fixes 4 cores for load normalisation + thresholds. */
const CORES = 4;

/* Round to n decimals (mock display precision only). */
function round(x, n) {
  const p = 10 ** n;

  return Math.round(x * p) / p;
}

/* A configured speed wins over the detected link (the NIC is often faster than
 * the uplink behind it); then sysfs, then 1 GbE. Full duplex: each direction
 * gets the whole link. */
function netRefBps(maxMbit, detectedMbit) {
  let mbit = maxMbit > 0 ? maxMbit : detectedMbit;
  if (!(mbit > 0))
    mbit = 1000;
  return mbit * 125000.0; /* Mbit/s -> bytes/s */
}

function netConvert(bps, unit, refBps) {
  if (unit && unit[0] === '%')
    return refBps > 0 ? (bps * 100.0) / refBps : 0.0;

  return bps;
}

function tempConvert(celsius, unit, ref) {
  if (!unit)
    return celsius;

  if (unit[0] === 'f')
    return (celsius * 9.0) / 5.0 + 32.0;

  if (unit[0] === '%')
    return ref > 0 ? (celsius * 100.0) / ref : celsius;

  return celsius;
}

/* 100% reference for temp percent mode: the sysfs critical trip point when
 * present (non-null), else the configured fallback. Mirrors the C server. */
function tempRef(crit, fallback) {
  return crit != null ? crit : fallback;
}

function loadConvert(load, cores, unit) {
  if (unit && unit[0] === '%' && cores > 0)
    return (load * 100.0) / cores;

  return load;
}

/* Mirror config_has(): absent list => show-all, else must contain the name. */
function configHas(list, name) {
  return list == null || (Array.isArray(list) && list.includes(name));
}

/* /api/current in CARD units. Mirrors json.c json_serialize_current(): a metric
 * group is emitted only when its card is configured, and [warn, crit] thresholds
 * are computed server-side. */
function toCurrent(raw, f, fallback, netMaxSpeed = 0) {
  const m = raw;
  const cards = f.cards;
  const lu = String(f.cpu_load_card_unit);
  const mu = String(f.mem_card_unit);
  const du = String(f.disk_card_unit);
  const nu = String(f.net_card_unit);
  const tu = String(f.temp_card_unit);

  const out = { timestamp: raw.timestamp };

  if (configHas(cards, 'cpu_load')) {
    out.load_1m = round(loadConvert(m.load_1m, CORES, lu), 2);
    out.load_5m = round(loadConvert(m.load_5m, CORES, lu), 2);
    out.load_15m = round(loadConvert(m.load_15m, CORES, lu), 2);
  }

  if (configHas(cards, 'cpu_usage')) {
    out.cpu_user_percent = round(m.cpu_user, 1);
    out.cpu_system_percent = round(m.cpu_system, 1);
    out.cpu_idle_percent = round(m.cpu_idle, 1);
  }

  if (configHas(cards, 'memory')) {
    if (mu[0] !== '%') {
      out.mem_used = round(m.mem_used_mb, 2);
      out.mem_available = round(m.mem_avail_mb, 2);
      out.mem_total = round(m.mem_total_mb, 2);
    }

    out.mem_percent = round(m.mem_percent, 1);
  }

  if (configHas(cards, 'disk')) {
    if (du[0] !== '%') {
      out.disk_used = round(m.disk_used_gb, 2);
      out.disk_total = round(m.disk_total_gb, 2);
      out.disk_free = round(m.disk_free_gb, 2);
    }

    out.disk_percent = round(m.disk_percent, 1);
  }

  const tc = m.temp_c;
  const crit = m.temp_critical_c;
  const ref = tempRef(crit, fallback);
  const nref = netRefBps(netMaxSpeed, m.net_speed_mbit);

  if (configHas(cards, 'temp')) {
    out.temp = tc != null ? round(tempConvert(tc, tu, ref), 1) : null;
  }

  if (configHas(cards, 'net')) {
    out.net_rx = round(netConvert(m.net_rx_bps, nu, nref), 2);
    out.net_tx = round(netConvert(m.net_tx_bps, nu, nref), 2);
  }

  if (configHas(cards, 'uptime')) {
    out.uptime_seconds = raw.uptime;
  }

  /* Thresholds, computed server-side ([warn, crit]). */
  if (configHas(cards, 'cpu_load')) {
    out.thresh_load = lu[0] === '%' ? [70.0, 90.0] : [0.75 * CORES, 1.0 * CORES];
  }

  if (configHas(cards, 'cpu_usage'))
    out.thresh_cpu = [70.0, 90.0];

  if (configHas(cards, 'memory'))
    out.thresh_mem = [70.0, 90.0];

  if (configHas(cards, 'disk'))
    out.thresh_disk = [80.0, 90.0];

  if (configHas(cards, 'temp')) {
    const base = crit != null ? [crit - 20.0, crit - 10.0] : [70.0, 80.0];

    out.thresh_temp = [
      round(tempConvert(base[0], tu, ref), 4),
      round(tempConvert(base[1], tu, ref), 4),
    ];
  }

  if (configHas(cards, 'net')) {
    out.thresh_net = [
      round(netConvert(0.85 * nref, nu, nref), 4),
      round(netConvert(0.98 * nref, nu, nref), 4),
    ];
  }

  return out;
}

/* Mirror of unit_step + json_serialize_units in the C server. */
const LADDERS = {
  mem:   { sym: ['MB', 'GB', 'TB'], mul: 1, factor: 1024, base: 0 },
  disk:  { sym: ['GB', 'TB', 'PB'], mul: 1, factor: 1024, base: 0 },
  bytes: { sym: ['KB/s', 'MB/s', 'GB/s'], mul: 1, factor: 1024, base: 1 },
  bits:  { sym: ['Kbps', 'Mbps', 'Gbps'], mul: 8, factor: 1000, base: 1 },
};
const DIGITS_MAX = 9999;

function unitStep(max, steps, factor) {
  if (!(max > 0)) return -1;
  let i = 0;
  while (i < steps - 1 && max / Math.pow(factor, i) > DIGITS_MAX) i++;
  return i;
}

function unitFor(max, kind) {
  const l = LADDERS[kind];
  const i = unitStep((max * l.mul) / Math.pow(l.factor, l.base), l.sym.length, l.factor);
  return i < 0 ? null : { sym: l.sym[i], mul: l.mul, div: Math.pow(l.factor, i + l.base) };
}

function unitsFor(raws, f) {
  const out = {};
  const peak = (rows, ...keys) => rows.reduce((m, r) =>
    keys.reduce((n, k) => (r[k] != null && r[k] > n ? r[k] : n), m), 0);
  /* Like the C server: a row with no delta yet carries noise, not throughput. */
  const withNet = raws.filter((r) => r.net_valid !== 0);

  if (String(f.mem_chart_unit)[0] !== '%')
    out.mem = unitFor(peak(raws, 'mem_used_mb', 'mem_avail_mb'), 'mem');
  if (String(f.disk_chart_unit)[0] !== '%')
    out.disk = unitFor(peak(raws, 'disk_used_gb', 'disk_free_gb'), 'disk');
  if (String(f.net_chart_unit)[0] !== '%') {
    out.net = unitFor(peak(withNet, 'net_rx_bps', 'net_tx_bps'),
      String(f.net_chart_unit) === 'bits' ? 'bits' : 'bytes');
  }
  Object.keys(out).forEach((k) => { if (!out[k]) delete out[k]; });
  return out;
}

/* Serialize a raw point for /api/metrics using the CHART units (short keys). */
function toPoint(raw, f, fallback, netMaxSpeed = 0) {
  const m = raw;
  const lu = String(f.cpu_load_chart_unit);
  const mu = String(f.mem_chart_unit);
  const du = String(f.disk_chart_unit);
  const nu = String(f.net_chart_unit);
  const tu = String(f.temp_chart_unit);

  const out = {
    t: raw.t,
    l1: round(loadConvert(m.load_1m, CORES, lu), 2),
    l5: round(loadConvert(m.load_5m, CORES, lu), 2),
    l15: round(loadConvert(m.load_15m, CORES, lu), 2),
    cu: round(m.cpu_user, 1),
    cs: round(m.cpu_system, 1),
    ci: round(m.cpu_idle, 1),
  };

  if (mu[0] !== '%') {
    out.mu = round(m.mem_used_mb, 2);
    out.ma = round(m.mem_avail_mb, 2);
    out.mt = round(m.mem_total_mb, 2);
  }

  out.mp = round(m.mem_percent, 1);

  if (du[0] !== '%') {
    out.du = round(m.disk_used_gb, 2);
    out.dt = round(m.disk_total_gb, 2);
    out.df = round(m.disk_free_gb, 2);
  }

  out.dp = round(m.disk_percent, 1);

  const tc = m.temp_c;
  const ref = tempRef(m.temp_critical_c, fallback);

  if (configHas(f.charts, 'temp')) {
    out.tp = tc != null ? round(tempConvert(tc, tu, ref), 1) : null;
  }

  const nref = netRefBps(netMaxSpeed, m.net_speed_mbit);

  out.nr = round(netConvert(m.net_rx_bps, nu, nref), 2);
  out.nt = round(netConvert(m.net_tx_bps, nu, nref), 2);
  out.up = raw.uptime;

  return out;
}

module.exports = {
  netConvert,
  netRefBps,
  tempConvert,
  tempRef,
  loadConvert,
  toCurrent,
  toPoint,
  unitsFor,
  unitStep,
};
